README.md
<!--
 * @Author: zhupengfei
 * @Date: 2021-09-08 15:07:05
 * @LastEditTime: 2021-09-16 15:43:42
 * @LastEditors: zhupengfei
 * @Description:
 * @FilePath: /cocos-build/README.md
-->

# Cocos Creator Build

This action provides the following functionality for Github Actions users:

- automate building for cocos creator project
- configuring cocos_download_url, cocos-version, cocos_type, project_path, platform, build_path for building
- only run on mac

| inputs             | required | description                 | default                                                                |
| ------------------ | -------- | --------------------------- | ---------------------------------------------------------------------- |
| cocos_download_url | false    | cocos creator download urls | https://creator-api.cocos.com/api/cocoshub/editor_version_list?lang=zh |
| cocos_version      | false    | cocos creator version       | 0.0.0                                                                  |
| cocos_type         | false    | cocos creator type(2d/3d)   | 2d                                                                     |
| project_path       | False    | project path for build      | ./                                                                     |
| platform           | false    | platform to build           | web-mobile                                                             |
| build_path         | false    | build path                  | ./build                                                                |

## Usage

See [action.yml]()

**Basic**

```
steps:
    - uses: actions/checkout@v2
    - name: Cocos Creator Build
    uses: miggene/cocos-build@v1.2.3
    with:
        cocos_version: 2.4.5
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.2.3](https://github.com/miggene/cocos-build/compare/v1.2.2...v1.2.3) (2021-09-16)


### Bug Fixes

* fix upload artifact config path ([13d5b0e](https://github.com/miggene/cocos-build/commit/13d5b0e7cd2632051e26aa50ac2f66cc67ad164a))

### [1.2.2](https://github.com/miggene/cocos-build/compare/v1.2.1...v1.2.2) (2021-09-16)

### [1.2.1](https://github.com/miggene/cocos-build/compare/v1.2.0...v1.2.1) (2021-09-16)

## [1.2.0](https://github.com/miggene/cocos-build/compare/v1.1.3...v1.2.0) (2021-09-16)


### Features

* add upload artifact feature ([5f8aa27](https://github.com/miggene/cocos-build/commit/5f8aa279ffb115a53bbc9edb32991f6797b97a03))

### [1.1.3](https://github.com/miggene/cocos-build/compare/v1.1.2...v1.1.3) (2021-09-16)

### [1.1.2](https://github.com/miggene/cocos-build/compare/v1.1.1...v1.1.2) (2021-09-16)

### [1.1.1](https://github.com/miggene/cocos-build/compare/v1.1.0...v1.1.1) (2021-09-15)


### Bug Fixes

* add upload artifact ([86b62d1](https://github.com/miggene/cocos-build/commit/86b62d1ad4cbbd284c80b844d324a2c460423ad2))

## [1.1.0](https://github.com/miggene/cocos-build/compare/v1.0.2...v1.1.0) (2021-09-15)


### Features

* add platform for the build ([52a7cae](https://github.com/miggene/cocos-build/commit/52a7caefea78f971723dbee0f27690c9aa340f9d))

### [1.0.2](https://github.com/miggene/cocos-build/compare/v1.0.1...v1.0.2) (2021-09-15)

### [1.0.1](https://github.com/miggene/cocos-build/compare/v0.1.0...v1.0.1) (2021-09-15)

## [0.1.0](https://github.com/miggene/cocos-build/compare/v0.0.19...v0.1.0) (2021-09-15)


### ⚠ BREAKING CHANGES

* **readme:** first release

### Features

* **readme:** a descript for this tool ([8aae056](https://github.com/miggene/cocos-build/commit/8aae0560936e62f13b640cc202ced19f4d1a9d36))

### [0.0.19](https://github.com/miggene/cocos-build/compare/v0.0.18...v0.0.19) (2021-09-15)

### [0.0.18](https://github.com/miggene/cocos-build/compare/v0.0.17...v0.0.18) (2021-09-15)

### [0.0.17](https://github.com/miggene/cocos-build/compare/v0.0.16...v0.0.17) (2021-09-15)

### [0.0.16](https://github.com/miggene/cocos-build/compare/v0.0.15...v0.0.16) (2021-09-15)

### [0.0.15](https://github.com/miggene/cocos-build/compare/v0.0.14...v0.0.15) (2021-09-15)

### [0.0.14](https://github.com/miggene/cocos-build/compare/v0.0.13...v0.0.14) (2021-09-15)

### [0.0.13](https://github.com/miggene/cocos-build/compare/v0.0.12...v0.0.13) (2021-09-15)

### [0.0.12](https://github.com/miggene/cocos-build/compare/v0.0.11...v0.0.12) (2021-09-15)

### [0.0.11](https://github.com/miggene/cocos-build/compare/v0.0.10...v0.0.11) (2021-09-15)

### [0.0.10](https://github.com/miggene/cocos-build/compare/v0.0.9...v0.0.10) (2021-09-15)

### [0.0.9](https://github.com/miggene/cocos-build/compare/v0.0.8...v0.0.9) (2021-09-15)

### [0.0.8](https://github.com/miggene/cocos-build/compare/v0.0.7...v0.0.8) (2021-09-14)

### [0.0.7](https://github.com/miggene/cocos-build/compare/v0.0.3...v0.0.7) (2021-09-14)

### [0.0.6](https://github.com/miggene/cocos-build/compare/v0.0.3...v0.0.6) (2021-09-14)

### [0.0.5](https://github.com/miggene/cocos-build/compare/v0.0.4...v0.0.5) (2021-09-14)

### [0.0.4](https://github.com/miggene/cocos-build/compare/v0.0.3...v0.0.4) (2021-09-14)

### [0.0.3](https://github.com/miggene/cocos-build/compare/v0.0.2...v0.0.3) (2021-09-14)

### [0.0.2](https://github.com/miggene/cocos-build/compare/v0.0.1...v0.0.2) (2021-09-14)


__tests__/main.test.ts
import {wait} from '../src/wait'
import * as process from 'process'
import * as cp from 'child_process'
import * as path from 'path'
import {expect, test} from '@jest/globals'

test('throws invalid number', async () => {
  const input = parseInt('foo', 10)
  await expect(wait(input)).rejects.toThrow('milliseconds not a number')
})

// test('wait 500 ms', async () => {
//   const start = new Date()
//   await wait(500)
//   const end = new Date()
//   var delta = Math.abs(end.getTime() - start.getTime())
//   expect(delta).toBeGreaterThan(450)
// })

// shows how the runner will run a javascript action with env / stdout protocol
// test('test runs', () => {
//   // process.env['INPUT_MILLISECONDS'] = '500'
//   process.env['INPUT_COCOS_DOWNLOAD_URL'] =
//     'https://creator-api.cocos.com/api/cocoshub/editor_version_list?lang=zh'
//   process.env['INPUT_COCOS_VERSION'] = '0.0.0'
//   process.env['INPUT_COCOS_TYPE'] = '2d'
//   process.env['INPUT_PROJECT_PATH'] = './'
//   const np = process.execPath
//   const ip = path.join(__dirname, '..', 'lib', 'main.js')
//   const options: cp.ExecFileSyncOptions = {
//     env: process.env
//   }
//   console.log(cp.execFileSync(np, [ip], options).toString())
// })


src/main.ts
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-explicit-any */
/*
 * @Author: zhupengfei
 * @Date: 2021-09-08 15:07:05
 * @LastEditTime: 2021-09-16 15:43:20
 * @LastEditors: zhupengfei
 * @Description:
 * @FilePath: /cocos-build/src/main.ts
 */
import * as core from '@actions/core'
import axios from 'axios'
import {exec} from '@actions/exec'
import {downloadTool, extractZip} from '@actions/tool-cache'
import * as artifact from '@actions/artifact'
import * as glob from '@actions/glob'

// import {wait} from './wait'

type CCDownloadType = {version: string; darwin: string; win32: string}

async function run(): Promise<void> {
  try {
    const downloadUrls = core.getInput('cocos_download_url')
    const cocosVersion = core.getInput('cocos_version')
    const cocosType = core.getInput('cocos_type')
    const projectPath = core.getInput('project_path')
    const platform = core.getInput('platform')
    const buildPath = core.getInput('build_path')
    try {
      const {data} = await (await axios.get(downloadUrls)).data
      const urlList = data[cocosType] as CCDownloadType[]
      const {version, darwin} =
        cocosVersion === '0.0.0'
          ? urlList[0]
          : urlList.find(value => {
              return value.version === cocosVersion
            })!
      const ccZipPath = await downloadTool(
        darwin,
        `CocosCreator_V${version}.zip`
      )
      await extractZip(`${ccZipPath}`, './')
      await exec(`open ./CocosCreator.app`)
      await exec(
        `./CocosCreator.app/Contents/MacOS/CocosCreator --path ${projectPath} --build "platform=${platform};buildPath=${buildPath}"`
      )
      const artifactClient = artifact.create()
      const artifactName = 'cocos-build-package'
      const patterns = `${buildPath}/${platform}`
      const globber = await glob.create(patterns)
      const files = await globber.glob()
      console.log('files :>> ', files)

      const rootDirectory = `${buildPath}`
      // const options = {
      //   continueOnError: true
      // }

      const uploadResult = await artifactClient.uploadArtifact(
        artifactName,
        files,
        rootDirectory
        // options
      )
      console.log('uploadResult :>> ', uploadResult)
    } catch (error) {
      core.error(error as string)
    }
  } catch (error: any) {
    core.setFailed(error.message)
  }
}

run()
